File: src/173. Binary Search Tree Iterator.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */

export class BSTIterator {
  /**
   * @param {TreeNode} root
   */
  constructor(root) {
    this.stack = []
    this._collectLeftNodes(root)
  }
  /**
   * @return {number}
   */
  next() {
    const node = this.stack.pop()
    if (node.right) this._collectLeftNodes(node.right)
    return node.val
  }
  /**
   * @return {boolean}
   */
  hasNext() {
    return !!this.stack.length
  }
  /**
   * @return {undefined}
   */
  _collectLeftNodes(node) {
    let current = node
    while (current) {
      this.stack.push(current)
      current = current.left
    }
  }
}

/**
 * Your BSTIterator object will be instantiated and called as such:
 * var obj = new BSTIterator(root)
 * var param_1 = obj.next()
 * var param_2 = obj.hasNext()
 */

File: src/173. Binary Search Tree Iterator.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class BSTIterator {
private:
    stack<TreeNode*> s;

public:
    BSTIterator(TreeNode* node) {
        while (node) {
            s.push(node);
            node = node->left;
        }
    }

    int next() {
        TreeNode* node = s.top();
        s.pop();

        const int val = node->val;
        node = node->right;
        while (node) {
            s.push(node);
            node = node->left;
        }

        return val;
    }

    bool hasNext() {
        return !s.empty();
    }
};

/**
 * Your BSTIterator object will be instantiated and called as such:
 * BSTIterator* obj = new BSTIterator(root);
 * int param_1 = obj->next();
 * bool param_2 = obj->hasNext();
 */